[ESP32] Remove dummy diagnostics log files support from temperature-measurement-app

pimpalemahesh opened the pull request:
#### Summary
Currently in temperature-measurement-app diagnostic-logs-delegate uses dummy log files. Default should use diagnostic-storage from esp32_diagnostic_trace tracing backend.

#### Testing

Build temperature-measurement-app for esp32 platform with enabling `CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE=y` and `CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE=n`
Get expected result in both cases.

pimpalemahesh committed:
Remove dummy diagnostics log file support from temperature-measurement-app

pimpalemahesh committed:
return unsupported feature if diagnostic trace not enabled

pimpalemahesh committed:
Move diagnostic delegate to platform common code

pimpalemahesh committed:
Move retreive diagnostic buffer

pimpalemahesh committed:
Move diagnostic backend initalization logic to delegate

pimpalemahesh committed:
Free allocated memory

dhrishi commented on examples/platform/esp32/diagnostics/diagnostic-logs-provider-delegate-impl.cpp:
```diff
@@ -31,7 +32,6 @@ LogProvider::CrashLogContext LogProvider::sCrashLogContext;
 
 #ifdef CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE
```
Is this option required here? Only for the `sReadEntries` variable. Instead should these files be guarded entirely with this option?

pimpalemahesh committed:
Add readme to enable diagnostics in app
- remove `CONFIG_ENABLE_ESP_DIAGNOSTICS_TRACE` option check from
  delegate files

pimpalemahesh committed:
Restyled by prettier-markdown

dhrishi commented:
@andy31415 @bzbarsky-apple Can you please take a look

pimpalemahesh committed:
Add init param for Log provider

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=39587) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=39587) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pimpalemahesh<br/>:x: esp<br/><hr/>**esp** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=39587) it.</sub>
